### PR TITLE
libsystem

### PR DESCRIPTION
**Pull request by arcnmx**
This is a WIP, looking for feedback on the overall concept/design mostly at this point... Probably requires an RFC to go through with properly?

tl;dr I ripped everything platform-specific out of `libstd` and implemented them behind a new `libsystem` platform-agnostic API.

Goals:
- Clearly define a platform-independent API that provides all functionality `libstd` needs without exposing any implementation details.
- Make porting `libstd` to new platforms straightforward by isolating all functionality that needs to be provided.
  - bare metal, kernels, microcontrollers, etc.
  - static syscall `libstd` that doesn't depend on any C libraries
- Cut the circular dependencies between `libstd/sys` and `libstd`
- Remove all `#[cfg(target_os..., unix, etc)]` branches in `libstd`.
- Remove `libstd`'s dependency on `liblibc`
- Implement a `bind` fallback `libsystem` implementation entirely made of `extern fn`s in order to...
  1. Have a reference implementation that exposes nothing about itself to ensure no implementation details are being relied upon.
  2. Allow libstd to be used raw without any dependency at all, while lazily linking in any functionality that does end up being necessary.
- Migrate as much as possible away from `rust_builtin` (low prio, but would be nice)

Also in this branch:
- Cargo.toml for all of rustc and its stdlib. Not incredibly useful compared to the [ongoing cargo-build branch](https://github.com/alexcrichton/rust/tree/cargo-build), but I like it as a quick way to build a `libstd` sysroot and the rustc compiler without having to deal with the multi-stage build system.
- target_family flexible target configuration: 836daa811de287f15dce1da7db1b202cbb66b8fc - the whole world isn't just unix and windows. This probably belongs as a standalone PR?
- `liblibc` changes for a generic target_family. Probably useless, better to just kill `liblibc` as a dependency from everything except `libsystem`.

### CODE REVIEW

**Comment by eefriedman** [comment]
I'd suggest starting a thread on internals.rust-lang.org discussing what your goal is, and what steps are necessary to reach it.  This PR is too big, and it implicitly makes a bunch of decisions which need to be separately discussed.

A few other comments:
- it isn't clear what you're trying to do with `target_family`; making it open-ended doesn't seem useful at first glance.
- It's probably worth reading https://github.com/rust-lang/rfcs/pull/1291 if you're going to suggest changes to libc.
- Does using libstd actually make sense in your environment?  Environments without a C runtime can use libcore (https://github.com/rust-lang/rust/issues/27701).

**Comment by alexcrichton** [comment]
Thanks for the PR! This is a pretty significant change, however, and it is currently requires that PRs of this magnitude go through at least an RFC to discuss the design with the broader community. There's a number of changes here which need to go through that process, so I'm going to close this for now. If you need any help writing an RFC though please feel free to let me know!

**Comment by arcnmx** [comment]
@alexcrichton @eefriedman I'll start with an internals thread then. Could you describe which parts need separate discussion? I only really see one item, which is "make `libstd::sys` self-contained and ensure platform-dependent behaviour only resides there". The rest is implementation details, mainly a question of whether to pull it out into a separate crate or leave it in libstd.

**Comment by alexcrichton** [comment]
I would personally want to discuss some of the high-level goals here as well as the general architecture for the structure of the standard library and crates beneath, I'm sure I'll have opinions! This also touches on other topics like the liblibc stabilization, libcore stabilization, bootstrapping with cargo, etc, that'd be good to make sure we're all aligned on before taking such a large step forward.

Regardless though let's continue discussion on the [internals thread you made](https://internals.rust-lang.org/t/libsystem-or-the-great-libstd-refactor/2765)!
